File: src/Samples/CSharpCodeSamples/AsynchronousExceptionsExamples.cs
﻿namespace CSharpCodeSamples
{
    using System;
    using System.Threading.Tasks;
    using Rackspace.Threading;

    public class AsynchronousExceptionsExamples
    {
        public void ExceptionPriorToTaskCreation()
        {
            #region ExceptionPriorToTaskCreation
            try
            {
                Task myTask = SomeOperationAsync();
            }
            catch (ArgumentException ex)
            {
                // ex was thrown directly by SomeOperationAsync. This cannot occur if SomeOperationAsync is an async
                // function (§10.15 - C# Language Specification Version 5.0).
            }
            #endregion
        }

        public void ExceptionDuringTaskExecution()
        {
            #region ExceptionDuringTaskExecution
            try
            {
                Task myTask = SomeOperationAsync();
                myTask.Wait();
            }
            catch (AggregateException wrapperEx)
            {
                ArgumentException ex = wrapperEx.InnerException as ArgumentException;
                if (ex == null)
                    throw;

                // ex was thrown during the asynchronous portion of SomeOperationAsync. This is always the case if
                // SomeOperationAsync is an async function (§10.15 - C# Language Specification Version 5.0).
            }
            #endregion
        }

        public void AsynchronousMethodAsContinuation()
        {
            #region AsynchronousMethodAsContinuation
            // original asynchronous method invocation
            Task task1 = SomeOperationAsync();

            // method invocation treated as a continuation
            Task task2 = CompletedTask.Default.Then(_ => SomeOperationAsync());
            #endregion
        }

        private static Task SomeOperationAsync()
        {
            throw new NotSupportedException();
        }
    }
}

File: src/Samples/VBCodeSamples/AsynchronousExceptionsExamples.vb
Imports System.Threading.Tasks
Imports Rackspace.Threading

Public Class AsynchronousExceptionsExamples

    Public Sub ExceptionPriorToTaskCreation()
        ' #Region ExceptionPriorToTaskCreation
        Try
            Dim myTask As Task = SomeOperationAsync()
        Catch ex As ArgumentException
            ' ex was thrown directly by SomeOperationAsync. If SomeOperationAsync Is marked with the `Async`
            ' keyword, then ex was thrown prior to the first use of the `Await` keyword within the implementation.
        End Try
        ' #End Region
    End Sub

    Public Sub ExceptionDuringTaskExecution()
        ' #Region ExceptionDuringTaskExecution
        Try
            Dim myTask As Task = SomeOperationAsync()
        Catch wrapperEx As AggregateException
            Dim ex = TryCast(wrapperEx.InnerException, ArgumentException)
            If ex Is Nothing Then
                Throw
            End If

            ' ex was thrown during the asynchronous portion of SomeOperationAsync. If SomeOperationAsync Is marked
            ' with the `Async` keyword, then ex was thrown after the first use of the `Await` keyword within the
            ' method.
        End Try
        ' #End Region
    End Sub

    Public Sub AsynchronousMethodAsContinuation()
        ' #Region AsynchronousMethodAsContinuation
        ' original asynchronous method invocation
        Dim task1 = SomeOperationAsync()

        ' method invocation treated as a continuation
        Dim task2 = CompletedTask.Default.Then(Function(task) SomeOperationAsync())
        ' #End Region
    End Sub

    Public Function SomeOperationAsync() As Task
        Throw New NotSupportedException()
    End Function

End Class
